src/ctia/stores/es/query.clj

```diff
@@ -53,3 +55,76 @@
+    {:default_operator s/Str
+     :fields [s/Str]})))
+
+(defn- searchable-fields-map-impl
+  ([m] (searchable-fields-map-impl [] m))
+  ([path m]
+   (if (and (map? m) (not= (:type m) "keyword"))
+     (let [step-fn (fn [k]
+                     (let [path (if (= k :properties) path (conj path k))]
+                       (searchable-fields-map-impl path (m k))))]
+       (mapcat step-fn (keys m)))
+     (let [fields (reduce-kv
+                   (fn [acc field {:keys [type]}]
+                     (if (= type "text")
+                       (conj acc field)
+                       acc))
+                   []
+                   (:fields m))]
+       (map #(conj [path] %) fields)))))
+
+(s/defn searchable-fields-map :- {s/Str s/Str}
+  "Walks through entity's mapping properties and finds all fields with secondary searchable token.
+Returns a map where key is path to a field, and value - path to the nested text token."
+  [properties :- (s/maybe {s/Keyword s/Any})]
+  (into {}
+        (map (fn [[path field]]
+               (let [path (str/join "." (map name path))]
+                 [path
+                  (str/join "." [path (name field)])])))
+        (searchable-fields-map-impl properties)))
+
+(s/defschema ESConnStateProps
+  (st/optional-keys
+   {:config {s/Any s/Any}
+    :props {s/Keyword s/Any}
+    :index s/Any
+    :conn s/Any
+    :services s/Any}))
+
+(s/defn rename-search-fields :- (s/maybe {s/Keyword [s/Str]})
+  "Automatically translates keyword fields to use underlying text field.
+
+   ES doesn't like when different types of tokens get used in the same query. To deal with
+   that, we create a nested field of type 'text', see:
+   `ctia.stores.es.mapping/searchable-token`. This should be opaque - caller shouldn't
+   have to explicitly instruct API to direct query to the nested field."
+  [es-conn-state :- ESConnStateProps
+   fields :- (s/maybe [s/Any])]
+  (let [properties (some-> es-conn-state :config :mappings first second :properties)
+        mapping (searchable-fields-map properties)]
+    (when (seq fields)
+      {:fields
+       (mapv (comp #(get mapping % %) name) fields)})))
+
+(s/defn refine-full-text-query-parts :- [{s/Keyword ESQFullTextQuery}]
+  [es-conn-state :- ESConnStateProps
+   full-text-terms :- [FullTextQuery]]
+  (let [{{:keys [default_operator]} :props} es-conn-state
+        term->es-query-part (fn [{:keys [query_mode fields] :as text-query}]
+                              (hash-map
+                               (or query_mode :query_string)
+                               (-> text-query
+                                   (dissoc :query_mode)
+                                   (merge
+                                    (when (and default_operator
+                                               (not= query_mode :multi_match))
+                                      {:default_operator default_operator})
+                                    (rename-search-fields es-conn-state fields)))))]
```

Comment:
We need to add the condition of the feature flag and turn it on after the refresh of the mappings.

Reply:
Same flag or should I create new one 🤔 https://github.com/threatgrid/ctia/blob/master/src/ctia/http/routes/crud.clj#L336

Reply:
New flag for now https://github.com/threatgrid/ctia/blob/rewrite-map_for_searchable-tokens/src/ctia/stores/es/query.clj#L131